### PR TITLE
Output UTXOs to CSV file

### PR DESCRIPTION
**Pull request by hansieodendaal**
## Description
Added a command-line parameter to write all UTXOs to file in the clear: value, spending key, commitment, features.

## Motivation and Context
Needed to export UTXOs in the clear from the encrypted SQLite wallet database.

## How Has This Been Tested?
Tested with the console wallet using the command file.

## Checklist:

* [X] I'm merging against the `development` branch.
* [X] I have squashed my commits into a single commit.

### CODE REVIEW

File: applications/tari_console_wallet/src/automation/commands.rs [hunk 1]
```diff
@@ -471,8 +477,32 @@ pub async fn command_runner(
...
+                if parsed.args.is_empty() {
+                    for (i, utxo) in utxos.iter().enumerate() {
+                        println!("{}. Value: {} {}", i + 1, utxo.value, utxo.features);
+                    }
+                } else if let ParsedArgument::CSVFile(file) = parsed.args[0].clone() {
+                    let factory = PedersenCommitmentFactory::default();
+                    let file = File::create(file).map_err(|e| CommandError::File(e.to_string()))?;
+                    let mut csv_file = LineWriter::new(file);
+                    csv_file
+                        .write_all(b"\"#\",\"Value (uT)\",\"Spending Key\",\"Commitment\",\"Features\"\n")
```
Comment (sdbondi) [inline applications/tari_console_wallet/src/automation/commands.rs]: Might be a little neater to use `writeln!` and `r##"(your string here)"##` instead of escaping the quotes
```suggestion
                        writeln!(csv_file, r##""#","Value (uT)","Spending Key","Commitment","Features""##)
```

File: applications/tari_console_wallet/src/automation/commands.rs [hunk 2]
```diff
@@ -471,8 +477,32 @@ pub async fn command_runner(
...
+                            utxo.value.0,
+                            utxo.spending_key.to_hex(),
+                            utxo.as_transaction_input(&factory, OutputFeatures::default())
+                                .commitment
+                                .to_hex(),
+                            utxo.features
+                        );
+                        csv_file
+                            .write_all(line.as_bytes())
+                            .map_err(|e| CommandError::File(e.to_string()))?;
```
Comment (sdbondi) [inline applications/tari_console_wallet/src/automation/commands.rs]: Same as above - using writeln! is better because it means that you do not need to allocate an extra String on the heap (obviously, in this part of the code it doesn't matter at all but thought I'd mention it)
```suggestion
                        writeln!(
                            csv_file,
                            r##""{}","{}","{}","{}","{}""##,
                            i + 1,
                            utxo.value.0,
                            utxo.spending_key.to_hex(),
                            utxo.as_transaction_input(&factory, OutputFeatures::default())
                                .commitment
                                .to_hex(),
                            utxo.features
                        )
                            .map_err(|e| CommandError::File(e.to_string()))?;
```

File: applications/tari_console_wallet/src/automation/commands.rs [hunk 3]
```diff
@@ -48,6 +52,8 @@ use tokio::{
     time::{delay_for, timeout},
 };
```
Comment (sdbondi) [inline applications/tari_console_wallet/src/automation/commands.rs:56]:
```suggestion
```

File: applications/tari_console_wallet/src/automation/commands.rs [hunk 4]
```diff
@@ -48,6 +52,8 @@ use tokio::{
...
 
+use crate::automation::command_parser::{ParsedArgument, ParsedCommand};
+
```
Comment (sdbondi) [inline applications/tari_console_wallet/src/automation/commands.rs]:
```suggestion
```

File: applications/tari_console_wallet/src/automation/command_parser.rs [hunk 5]
```diff
@@ -216,6 +218,21 @@ fn parse_send_tari(mut args: SplitWhitespace) -> Result<Vec<ParsedArgument>, Par
...
+    let mut parsed_args = Vec::new();
+
+    // amount
+    if let Some(v) = args.next() {
+        if v == "csv_file" {
+            if let Some(v) = args.next() {
+                parsed_args.push(ParsedArgument::CSVFile(v.to_string()));
+            }
+        }
+    };
```
Comment (sdbondi) [inline applications/tari_console_wallet/src/automation/command_parser.rs]: This is a bit strange to put the csv_file literal as a command argument and it's looking more like we'll need a standard way to push command output to various streams (file, csv, stdout etc.)

For now perhaps `list-utxos csv:my-file.csv` or `list-utxos csv-file:my-file.csv`?

```suggestion
    let parse_args = args.next().and_then(|s| {
      let mut parts = s.split(":");
      // Return None if first part isn't "csv"
      parts.next().filter(|s| s == "csv")?;
      parts.next().map(|s|   ParsedArgument::CSVFile(v.to_string()))
     }).into_iter().collect();
```

**Review by sdbondi — COMMENTED** [review]
Looks good - few nitty comments

File: applications/tari_console_wallet/README.md [hunk 6]
```diff
@@ -72,15 +72,16 @@ Monitoring 1 sent transactions to Broadcast stage...
...
+`tari_console_wallet --command "export-utxos"`
+`tari_console_wallet --command "export-utxos csv_file <file name>"`
 
 example output:
 ```
-1. list-utxos
+1. export-utxos
 1. Value: 6000 µT OutputFeatures: Flags = (empty), Maturity = 0
 2. Value: 10000 µT OutputFeatures: Flags = (empty), Maturity = 0
 ...
```
Comment (SWvheerden) [inline applications/tari_console_wallet/README.md]: This is incomplete, 
Explain how the csv works etc. There is no mention of this in the help, only I  command example. 

The example should also be done properly. 


For example:
```
`tari_console_wallet --command "export-utxos"`
example output:

1. export-utxos
1. Value: 6000 µT OutputFeatures: Flags = (empty), Maturity = 0
2. Value: 10000 µT OutputFeatures: Flags = (empty), Maturity = 0
...

`tari_console_wallet --command "export-utxos csv_file <file name>"`
example output:
....
```

Currently reading the text, I don't know which one to give, to provide me that output.

File: applications/tari_console_wallet/src/automation/command_parser.rs [hunk 7]
```diff
@@ -302,4 +328,13 @@ fn test_parse_command() {
     } else {
         panic!("Parsed public key is not the same as provided.");
     }
+
+    let command_str = "export-utxos csv_file utxo_list.csv".to_string();
+    let parsed = parse_command(&command_str).unwrap();
+
+    if let ParsedArgument::CSVFileName(file) = parsed.args[0].clone() {
+        assert_eq!(file, "utxo_list.csv".to_string());
```
Comment (SWvheerden) [inline applications/tari_console_wallet/src/automation/command_parser.rs]: Why is this here?
This is not test code

Comment (hansieodendaal) [reply applications/tari_console_wallet/src/automation/command_parser.rs]: This one is inside a test.

File: applications/tari_console_wallet/src/automation/command_parser.rs (hunk 5 above)
Comment (hansieodendaal) [reply applications/tari_console_wallet/src/automation/command_parser.rs]: _Note: This has been changed since the review._

Personally, I prefer using s space over a colon as the separation character for the user format. I did change it slightly though to feel less strange.

From the command line:
`tari_console_wallet --command "export-utxos --csv-file my-file.csv"`
vs.
`tari_console_wallet --command "export-utxos csv_file my-file.csv"`

Using a command file:
`export-utxos --csv-file my-file.csv`
vs.
`export-utxos csv_file my-file.csv`

**Comment by hansieodendaal** [comment]
@sdbondi and @SWvheerden thank you for the reviews!

File: applications/tari_console_wallet/src/automation/command_parser.rs (hunk 5 above)
Comment (sdbondi) [reply applications/tari_console_wallet/src/automation/command_parser.rs]: yeah LGTM - my point was mostly that we should maybe have a standard for these commands. I do think that `--csv-file` makes it clearer that csv_file is an option for the command 👍🏻

**Review by sdbondi — DISMISSED** [review]
LGTM

File: applications/tari_console_wallet/src/automation/command_parser.rs (hunk 7 above)
Comment (SWvheerden) [reply applications/tari_console_wallet/src/automation/command_parser.rs:337]: This is not a unit test, right?

Comment (SWvheerden) [reply applications/tari_console_wallet/src/automation/command_parser.rs]: You are correct, this is a unit test. 
But is this not behind a cfg(test) flag?

Comment (delta1) [reply applications/tari_console_wallet/src/automation/command_parser.rs]: It's not behind a feature flag, obviously an oversight by the original author (me). It's not really that big a deal, but easy enough to add 

```rust
#[cfg(test)]
mod test {
```

Comment (hansieodendaal) [reply applications/tari_console_wallet/src/automation/command_parser.rs]: Done

**Review by stringhandler — APPROVED** [review]
